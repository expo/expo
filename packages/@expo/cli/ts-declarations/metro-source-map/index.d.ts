declare module 'metro-source-map' {
  //#region metro-source-map/src/source-map.js.flow

  type GeneratedCodeMapping = [number, number];
  type SourceMapping = [number, number, number, number];
  type SourceMappingWithName = [number, number, number, number, string];

  type HermesFunctionOffsets = { [key: number]: Readonly<Array<number>> };

  type FBSourcesArray = Readonly<Array<FBSourceMetadata | null>>;
  type FBSourceMetadata = [FBSourceFunctionMap | null];
  export type FBSourceFunctionMap = {
    names: Readonly<Array<string>>;
    mappings: string;
  };

  type FBSegmentMap = { [id: string]: MixedSourceMap };

  interface BasicSourceMap {
    file?: string;
    mappings: string;
    names: Array<string>;
    sourceRoot?: string;
    sources: Array<string>;
    sourcesContent?: Array<string | null>;
    version: number;
    x_facebook_offsets?: Array<number>;
    x_metro_module_paths?: Array<string>;
    x_facebook_sources?: FBSourcesArray;
    x_facebook_segments?: FBSegmentMap;
    x_hermes_function_offsets?: HermesFunctionOffsets;
  }

  interface IndexMapSection {
    map: IndexMap | BasicSourceMap;
    offset: {
      line: number;
      column: number;
    };
  }

  interface IndexMap {
    file?: string;
    mappings?: void; // avoids SourceMap being a disjoint union
    sections: Array<IndexMapSection>;
    version: number;
    x_facebook_offsets?: Array<number>;
    x_metro_module_paths?: Array<string>;
    x_facebook_sources?: FBSourcesArray;
    x_facebook_segments?: FBSegmentMap;
    x_hermes_function_offsets?: HermesFunctionOffsets;
  }

  type MixedSourceMap = IndexMap | BasicSourceMap;

  //#endregion

  //#region metro-source-map/src/composeSourceMaps.js.flow

  export function composeSourceMaps(maps: Readonly<Array<MixedSourceMap>>): MixedSourceMap;

  //#endregion

  //#region metro-source-map/src/composeSourceMaps.js.flow

  import type { Ast } from '@babel/core';

  type Context = { filename?: string };

  /**
   * Generate a map of source positions to function names. The names are meant to
   * describe the stack frame in an error trace and may contain more contextual
   * information than just the actual name of the function.
   *
   * The output is encoded for use in a source map. For details about the format,
   * see MappingEncoder below.
   */
  export function generateFunctionMap(ast: Ast, context?: Context): FBSourceFunctionMap;

  //#endregion
}
