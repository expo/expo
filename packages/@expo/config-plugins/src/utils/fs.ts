import fs from 'fs';
import path from 'path';

/** A basic function that copies a single file to another file location. */
export async function copyFilePathToPathAsync(src: string, dest: string): Promise<void> {
  const srcFile = await fs.promises.readFile(src);
  await fs.promises.mkdir(path.dirname(dest), { recursive: true });
  await fs.promises.writeFile(dest, srcFile);
}

/** Remove a single file (not directory). Returns `true` if a file was actually deleted. */
export function removeFile(filePath: string): boolean {
  try {
    fs.unlinkSync(filePath);
    return true;
  } catch (error: any) {
    // Skip if the remove did nothing.
    if (error.code === 'ENOENT') {
      return false;
    }
    throw error;
  }
}
