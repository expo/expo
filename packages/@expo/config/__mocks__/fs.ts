module.exports = require('memfs').fs;

// NOTE(cedric): workaround to also mock `node:fs`
jest.mock('node:fs', () => require('fs'));
