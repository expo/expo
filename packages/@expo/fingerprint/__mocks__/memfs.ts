const memfs = jest.requireActual('memfs');
const path = require('path');

/**
 * Further mock memfs `fs.promises` to ensure the parent directory exists.
 * We used to call `fs.promises.mkdtemp(path.join(os.tmpdir(), ...))`, so that we don't have to worry about os.tmpdir() exists.
 */
const origMkdtemp = memfs.fs.promises.mkdtemp;
memfs.fs.promises.mkdtemp = (
  prefix: string,
  options?: Parameters<typeof origMkdtemp>[1]
): ReturnType<typeof origMkdtemp> => {
  memfs.fs.mkdirSync(path.dirname(prefix), { recursive: true });
  return origMkdtemp(prefix, options);
};

module.exports = memfs;
