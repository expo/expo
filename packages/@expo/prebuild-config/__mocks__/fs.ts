const fs = require('memfs').fs;

// Add fs.promises.cp polyfill for memfs since it doesn't support it
fs.promises.cp = async (src: string, dest: string, options?: { recursive?: boolean }) => {
  const path = require('path');

  const srcStat = await fs.promises.stat(src);

  if (srcStat.isDirectory()) {
    await fs.promises.mkdir(dest, { recursive: true });
    const entries = await fs.promises.readdir(src, { withFileTypes: true });

    for (const entry of entries) {
      const srcPath = path.join(src, entry.name);
      const destPath = path.join(dest, entry.name);

      if (entry.isDirectory()) {
        await fs.promises.cp(srcPath, destPath, options);
      } else {
        await fs.promises.copyFile(srcPath, destPath);
      }
    }
  } else {
    await fs.promises.copyFile(src, dest);
  }
};

module.exports = fs;

jest.mock('node:fs', () => require('memfs').fs);
