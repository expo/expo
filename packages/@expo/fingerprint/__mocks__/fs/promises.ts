import { fs } from 'memfs';
import path from 'path';

/**
 * Further mock `fs.promises` to ensure the parent directory exists.
 * We used to call `fs.promises.mkdtemp(path.join(os.tmpdir(), ...))`, so that we don't have to worry about os.tmpdir() exists.
 */
const origMkdtemp = fs.promises.mkdtemp;
fs.promises.mkdtemp = (
  prefix: string,
  options?: Parameters<typeof origMkdtemp>[1]
): ReturnType<typeof origMkdtemp> => {
  fs.mkdirSync(path.dirname(prefix), { recursive: true });
  return origMkdtemp(prefix, options);
};

module.exports = fs.promises;
