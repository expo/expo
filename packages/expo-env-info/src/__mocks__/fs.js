import { fs, promises } from 'memfs';

const constants = jest.requireActual('fs').constants;

export { fs, promises, constants };
