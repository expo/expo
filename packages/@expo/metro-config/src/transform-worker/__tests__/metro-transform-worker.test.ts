/**
 * Copyright 2023-present 650 Industries (Expo). All rights reserved.
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * Fork of the Metro transformer worker, but with additional transforms moved to `babel-preset-expo` and modifications made for web support.
 * https://github.com/facebook/metro/blob/412771475c540b6f85d75d9dcd5a39a6e0753582/packages/metro-transform-worker/src/index.js#L1
 */

import { TraceMap, originalPositionFor, generatedPositionFor } from '@jridgewell/trace-mapping';
import { Buffer } from 'buffer';
import * as fs from 'fs';
import { vol } from 'memfs';
import { fromRawMappings } from 'metro-source-map';
import type { JsTransformerConfig, JsTransformOptions, JsOutput } from 'metro-transform-worker';
import * as path from 'path';

/** Converts source mappings from Metro to a “TraceMap”, which is similar to source-map’s SourceMapConsumer */
const toTraceMap = (output: JsOutput, contents: string) => {
  const map = fromRawMappings([output.data]).toMap();
  return new TraceMap({
    ...map,
    file: output.data.code,
    sources: [null],
    sourcesContent: [contents || null],
  });
};

const originalWarn = console.warn;

beforeEach(() => {
  console.warn = jest.fn();
});

afterEach(() => {
  console.warn = originalWarn;
});

jest
  .mock('metro-transform-worker/src/utils/getMinifier', () => () => ({ code, map, config }) => {
    const trimmed = config.output.comments ? code : code.replace('/*#__PURE__*/', '');
    return {
      code: trimmed.replace('arbitrary(code)', 'minified(code)'),
      map,
    };
  })
  .mock('metro-transform-plugins', () => ({
    ...jest.requireActual('metro-transform-plugins'),
    inlinePlugin: () => ({}),
    constantFoldingPlugin: () => ({}),
  }))
  .mock('metro-minify-terser');

const babelTransformerPath = require.resolve('@expo/metro-config/babel-transformer');

const transformerContents = jest.requireActual('fs').readFileSync(babelTransformerPath);

const HEADER_DEV =
  '__d(function (global, _$$_REQUIRE, _$$_IMPORT_DEFAULT, _$$_IMPORT_ALL, module, exports, _dependencyMap) {';
const HEADER_PROD = '__d(function (g, r, i, a, m, e, d) {';

// let fs: typeof import('fs');
let Transformer: typeof import('../metro-transform-worker');

const baseConfig: JsTransformerConfig = {
  allowOptionalDependencies: false,
  assetPlugins: [],
  assetRegistryPath: '',
  asyncRequireModulePath: 'asyncRequire',
  babelTransformerPath,
  dynamicDepsInPackages: 'reject',
  enableBabelRCLookup: false,
  enableBabelRuntime: true,
  globalPrefix: '',
  hermesParser: false,
  minifierConfig: { output: { comments: false } },
  minifierPath: 'minifyModulePath',
  optimizationSizeLimit: 100000,
  publicPath: '/assets',
  unstable_dependencyMapReservedName: null,
  unstable_compactOutput: false,
  unstable_disableModuleWrapping: false,
  unstable_disableNormalizePseudoGlobals: false,
  unstable_allowRequireContext: false,
  unstable_collectDependenciesPath: null,
};

const baseTransformOptions: JsTransformOptions = {
  dev: true,
  hot: false,
  inlinePlatform: false,
  inlineRequires: false,
  minify: false,
  platform: 'ios',
  type: 'module',
  unstable_transformProfile: 'default',
  customTransformOptions: {
    __proto__: null,
  },
};

jest.mock('fs');

beforeEach(() => {
  jest.resetModules();

  //   jest.mock('fs', () => new (require('metro-memory-fs'))());

  //   fs = require('fs');
  Transformer = require('../metro-transform-worker');

  vol.reset();

  fs.mkdirSync('/root/local', { recursive: true });
  fs.mkdirSync(path.dirname(babelTransformerPath), { recursive: true });
  fs.writeFileSync(babelTransformerPath, transformerContents);
});

it('transforms a simple script', async () => {
  const contents = 'someReallyArbitrary(code)';

  const result = await Transformer.transform(
    baseConfig,
    '/root',
    'local/file.js',
    Buffer.from(contents, 'utf8'),
    { ...baseTransformOptions, type: 'script' }
  );

  expect(result.output[0].type).toBe('js/script');
  expect(result.output[0].data.code).toBe(
    [
      '(function (global) {',
      '  someReallyArbitrary(code);',
      "})(typeof globalThis !== 'undefined' ? globalThis : typeof global !== 'undefined' ? global : typeof window !== 'undefined' ? window : this);",
    ].join('\n')
  );

  const trace = toTraceMap(result.output[0], contents);

  expect(generatedPositionFor(trace, { source: '', line: 1, column: 0 })).toMatchObject({
    line: 2,
    column: 2,
  });

  expect(originalPositionFor(trace, { line: 2, column: 2 })).toMatchObject({
    line: 1,
    column: 0,
    name: 'someReallyArbitrary',
  });

  // NOTE: If downgraded below @babel/generator@7.21.0, names will be missing here
  expect(originalPositionFor(trace, { line: 3, column: 10 })).toMatchObject({
    line: 1,
    column: 25,
    name: 'globalThis',
  });
  expect(originalPositionFor(trace, { line: 3, column: 59 })).toMatchObject({
    line: 1,
    column: 25,
    name: 'global',
  });
  expect(originalPositionFor(trace, { line: 3, column: 100 })).toMatchObject({
    line: 1,
    column: 25,
    name: 'window',
  });

  expect(result.output[0].data.functionMap).toMatchSnapshot();
  expect(result.dependencies).toEqual([]);
});

it('transforms a simple module', async () => {
  const contents = 'arbitrary(code)';

  const result = await Transformer.transform(
    baseConfig,
    '/root',
    'local/file.js',
    Buffer.from(contents, 'utf8'),
    baseTransformOptions
  );

  const trace = toTraceMap(result.output[0], contents);

  expect(generatedPositionFor(trace, { source: '', line: 1, column: 0 })).toMatchObject({
    line: 2,
    column: 2,
  });
  expect(generatedPositionFor(trace, { source: '', line: 1, column: 10 })).toMatchObject({
    line: 2,
    column: 12,
  });

  expect(originalPositionFor(trace, { line: 2, column: 2 })).toMatchObject({
    line: 1,
    column: 0,
    name: 'arbitrary',
  });
  expect(originalPositionFor(trace, { line: 2, column: 12 })).toMatchObject({
    line: 1,
    column: 10,
    name: 'code',
  });

  expect(result.output[0].type).toBe('js/module');
  expect(result.output[0].data.code).toBe([HEADER_DEV, '  arbitrary(code);', '});'].join('\n'));
  expect(result.output[0].data.functionMap).toMatchSnapshot();
  expect(result.dependencies).toEqual([]);
});

it('transforms a module with dependencies', async () => {
  const contents = [
    '"use strict";',
    'require("./a");',
    'arbitrary(code);',
    'const b = require("b");',
    'import c from "./c";',
  ].join('\n');

  const result = await Transformer.transform(
    baseConfig,
    '/root',
    'local/file.js',
    Buffer.from(contents, 'utf8'),
    baseTransformOptions
  );

  expect(result.output[0].type).toBe('js/module');
  expect(result.output[0].data.code).toBe(
    [
      HEADER_DEV,
      '  "use strict";',
      '',
      '  var _interopRequireDefault = _$$_REQUIRE(_dependencyMap[0], "@babel/runtime/helpers/interopRequireDefault");',
      '  var _c = _interopRequireDefault(_$$_REQUIRE(_dependencyMap[1], "./c"));',
      '  _$$_REQUIRE(_dependencyMap[2], "./a");',
      '  arbitrary(code);',
      '  var b = _$$_REQUIRE(_dependencyMap[3], "b");',
      '});',
    ].join('\n')
  );

  const trace = toTraceMap(result.output[0], contents);

  expect(
    generatedPositionFor(trace, { source: '', line: 2, column: 0 } /* require("./a") */)
  ).toMatchObject({ line: 6, column: 2 });

  expect(originalPositionFor(trace, { line: 7, column: 2 })).toMatchObject({
    line: 3,
    column: 0,
    name: 'arbitrary',
  });
  expect(originalPositionFor(trace, { line: 8, column: 6 })).toMatchObject({
    line: 4,
    column: 6,
    name: 'b',
  });
  expect(originalPositionFor(trace, { line: 8, column: 6 })).toMatchObject({
    line: 4,
    column: 6,
    name: 'b',
  });

  // NOTE: If downgraded below @babel/generator@7.21.0, names will be missing here
  expect(originalPositionFor(trace, { line: 5, column: 6 })).toMatchObject({
    line: 5,
    column: 0,
    name: '_c',
  });
  expect(originalPositionFor(trace, { line: 4, column: 6 })).toMatchObject({
    line: 1,
    column: 13,
    name: '_interopRequireDefault',
  });
  expect(originalPositionFor(trace, { line: 4, column: 31 })).toMatchObject({
    line: 1,
    column: 13,
    name: '_$$_REQUIRE',
  });
  expect(originalPositionFor(trace, { line: 4, column: 43 })).toMatchObject({
    line: 1,
    column: 13,
    name: '_dependencyMap',
  });

  expect(result.output[0].data.functionMap).toMatchSnapshot();

  expect(result.dependencies).toEqual([
    {
      data: expect.objectContaining({ asyncType: null }),
      name: '@babel/runtime/helpers/interopRequireDefault',
    },
    { data: expect.objectContaining({ asyncType: null }), name: './c' },
    { data: expect.objectContaining({ asyncType: null }), name: './a' },
    { data: expect.objectContaining({ asyncType: null }), name: 'b' },
  ]);
});

it('transforms an es module with asyncToGenerator', async () => {
  const contents = 'export async function test() {}';

  const result = await Transformer.transform(
    baseConfig,
    '/root',
    'local/file.js',
    Buffer.from(contents, 'utf8'),
    baseTransformOptions
  );

  expect(result.output[0].type).toBe('js/module');
  expect(result.output[0].data.code).toMatchSnapshot();

  const trace = toTraceMap(result.output[0], contents);

  expect(generatedPositionFor(trace, { source: '', line: 1, column: 12 })).toMatchObject({
    line: 12,
    column: 44,
  });

  expect(originalPositionFor(trace, { line: 8, column: 11 })).toMatchObject({
    line: 1,
    column: 22,
    name: 'test',
  });

  expect(result.output[0].data.functionMap).toMatchSnapshot();

  expect(result.dependencies).toEqual([
    {
      data: expect.objectContaining({ asyncType: null }),
      name: '@babel/runtime/helpers/interopRequireDefault',
    },
    {
      data: expect.objectContaining({ asyncType: null }),
      name: '@babel/runtime/helpers/asyncToGenerator',
    },
  ]);
});

it('transforms async generators', async () => {
  const result = await Transformer.transform(
    baseConfig,
    '/root',
    'local/file.js',
    Buffer.from('export async function* test() { yield "ok"; }', 'utf8'),
    baseTransformOptions
  );

  expect(result.output[0].data.code).toMatchSnapshot();
  expect(result.dependencies).toEqual([
    {
      data: expect.objectContaining({ asyncType: null }),
      name: '@babel/runtime/helpers/interopRequireDefault',
    },
    {
      data: expect.objectContaining({ asyncType: null }),
      name: '@babel/runtime/helpers/awaitAsyncGenerator',
    },
    {
      data: expect.objectContaining({ asyncType: null }),
      name: '@babel/runtime/helpers/wrapAsyncGenerator',
    },
  ]);
});

it('transforms import/export syntax when experimental flag is on', async () => {
  const contents = ['import c from "./c";'].join('\n');

  const result = await Transformer.transform(
    baseConfig,
    '/root',
    'local/file.js',
    Buffer.from(contents, 'utf8'),
    { ...baseTransformOptions, experimentalImportSupport: true }
  );

  expect(result.output[0].type).toBe('js/module');
  expect(result.output[0].data.code).toBe(
    [
      HEADER_DEV,
      '  "use strict";',
      '',
      '  var c = _$$_IMPORT_DEFAULT(_dependencyMap[0], "./c");',
      '});',
    ].join('\n')
  );

  const trace = toTraceMap(result.output[0], contents);

  expect(generatedPositionFor(trace, { source: '', line: 1, column: 7 })).toMatchObject({
    line: 4,
    column: 6,
  });

  expect(originalPositionFor(trace, { line: 4, column: 6 })).toMatchObject({
    line: 1,
    column: 7,
    name: 'c',
  });

  // NOTE: If downgraded below @babel/generator@7.21.0, names will be missing here
  expect(originalPositionFor(trace, { line: 4, column: 10 })).toMatchObject({
    line: 1,
    column: 8,
    name: '_$$_IMPORT_DEFAULT',
  });
  expect(originalPositionFor(trace, { line: 4, column: 29 })).toMatchObject({
    line: 1,
    column: 8,
    name: '_dependencyMap',
  });

  expect(result.output[0].data.functionMap).toMatchSnapshot();

  expect(result.dependencies).toEqual([
    {
      data: expect.objectContaining({
        asyncType: null,
      }),
      name: './c',
    },
  ]);
});

it('does not add "use strict" on non-modules', async () => {
  const result = await Transformer.transform(
    baseConfig,
    '/root',
    'node_modules/local/file.js',
    Buffer.from('module.exports = {};', 'utf8'),
    { ...baseTransformOptions, experimentalImportSupport: true }
  );

  expect(result.output[0].type).toBe('js/module');
  expect(result.output[0].data.code).toBe([HEADER_DEV, '  module.exports = {};', '});'].join('\n'));
});

it('preserves require() calls when module wrapping is disabled', async () => {
  const contents = ['require("./c");'].join('\n');

  const result = await Transformer.transform(
    {
      ...baseConfig,
      unstable_disableModuleWrapping: true,
    },
    '/root',
    'local/file.js',
    Buffer.from(contents, 'utf8'),
    baseTransformOptions
  );

  expect(result.output[0].type).toBe('js/module');
  expect(result.output[0].data.code).toBe('require("./c");');
});

it('reports filename when encountering unsupported dynamic dependency', async () => {
  const contents = ['require("./a");', 'let a = arbitrary(code);', 'const b = require(a);'].join(
    '\n'
  );

  try {
    await Transformer.transform(
      baseConfig,
      '/root',
      'local/file.js',
      Buffer.from(contents, 'utf8'),
      baseTransformOptions
    );
    throw new Error('should not reach this');
  } catch (error) {
    expect(error.message).toMatchSnapshot();
  }
});

it('supports dynamic dependencies from within `node_modules`', async () => {
  expect(
    (
      await Transformer.transform(
        {
          ...baseConfig,
          dynamicDepsInPackages: 'throwAtRuntime',
        },
        '/root',
        'node_modules/foo/bar.js',
        Buffer.from('require(foo.bar);', 'utf8'),
        baseTransformOptions
      )
    ).output[0].data.code
  ).toBe(
    [
      HEADER_DEV,
      '  (function (line) {',
      "    throw new Error('Dynamic require defined at line ' + line + '; not supported by Metro');",
      '  })(1);',
      '});',
    ].join('\n')
  );
});

it('minifies the code correctly', async () => {
  expect(
    (
      await Transformer.transform(
        baseConfig,
        '/root',
        'local/file.js',
        Buffer.from('arbitrary(code);', 'utf8'),
        { ...baseTransformOptions, minify: true }
      )
    ).output[0].data.code
  ).toBe([HEADER_PROD, '  minified(code);', '});'].join('\n'));
});

it('minifies a JSON file', async () => {
  expect(
    (
      await Transformer.transform(
        baseConfig,
        '/root',
        'local/file.json',
        Buffer.from('arbitrary(code);', 'utf8'),
        { ...baseTransformOptions, minify: true }
      )
    ).output[0].data.code
  ).toBe(
    [
      '__d(function(global, require, _importDefaultUnused, _importAllUnused, module, exports, _dependencyMapUnused) {',
      '  module.exports = minified(code);;',
      '});',
    ].join('\n')
  );
});

it('does not wrap a JSON file when disableModuleWrapping is enabled', async () => {
  expect(
    (
      await Transformer.transform(
        {
          ...baseConfig,
          unstable_disableModuleWrapping: true,
        },
        '/root',
        'local/file.json',
        Buffer.from('arbitrary(code);', 'utf8'),
        baseTransformOptions
      )
    ).output[0].data.code
  ).toBe('module.exports = arbitrary(code);;');
});

it('uses a reserved dependency map name and prevents it from being minified', async () => {
  const result = await Transformer.transform(
    { ...baseConfig, unstable_dependencyMapReservedName: 'THE_DEP_MAP' },
    '/root',
    'local/file.js',
    Buffer.from('arbitrary(code);', 'utf8'),
    { ...baseTransformOptions, dev: false, minify: true }
  );
  expect(result.output[0].data.code).toMatchInlineSnapshot(`
    "__d(function (g, r, i, a, m, e, THE_DEP_MAP) {
      minified(code);
    });"
  `);
});

it('throws if the reserved dependency map name appears in the input', async () => {
  await expect(
    Transformer.transform(
      { ...baseConfig, unstable_dependencyMapReservedName: 'THE_DEP_MAP' },
      '/root',
      'local/file.js',
      Buffer.from(
        'arbitrary(code); /* the code is not allowed to mention THE_DEP_MAP, even in a comment */',
        'utf8'
      ),
      { ...baseTransformOptions, dev: false, minify: true }
    )
  ).rejects.toThrowErrorMatchingInlineSnapshot(
    `"Source code contains the reserved string \`THE_DEP_MAP\` at character offset 55"`
  );
});

it('allows disabling the normalizePseudoGlobals pass when minifying', async () => {
  const result = await Transformer.transform(
    { ...baseConfig, unstable_disableNormalizePseudoGlobals: true },
    '/root',
    'local/file.js',
    Buffer.from('arbitrary(code);', 'utf8'),
    { ...baseTransformOptions, dev: false, minify: true }
  );
  expect(result.output[0].data.code).toMatchInlineSnapshot(`
    "__d(function (global, _$$_REQUIRE, _$$_IMPORT_DEFAULT, _$$_IMPORT_ALL, module, exports, _dependencyMap) {
      minified(code);
    });"
  `);
});

it('allows emitting compact code when not minifying', async () => {
  const result = await Transformer.transform(
    { ...baseConfig, unstable_compactOutput: true },
    '/root',
    'local/file.js',
    Buffer.from('arbitrary(code);', 'utf8'),
    { ...baseTransformOptions, dev: false, minify: false }
  );
  expect(result.output[0].data.code).toMatchInlineSnapshot(
    `"__d(function(global,_$$_REQUIRE,_$$_IMPORT_DEFAULT,_$$_IMPORT_ALL,module,exports,_dependencyMap){arbitrary(code);});"`
  );
});

it('skips minification in Hermes stable transform profile', async () => {
  const result = await Transformer.transform(
    baseConfig,
    '/root',
    'local/file.js',
    Buffer.from('arbitrary(code);', 'utf8'),
    {
      ...baseTransformOptions,
      dev: false,
      minify: true,
      unstable_transformProfile: 'hermes-canary',
      customTransformOptions: { __proto__: null, bytecode: true },
    }
  );
  expect(result.output[0].data.code).toMatchInlineSnapshot(`
    "__d(function (global, _$$_REQUIRE, _$$_IMPORT_DEFAULT, _$$_IMPORT_ALL, module, exports, _dependencyMap) {
      arbitrary(code);
    });"
  `);
});

it('skips minification in Hermes canary transform profile', async () => {
  const result = await Transformer.transform(
    baseConfig,
    '/root',
    'local/file.js',
    Buffer.from('arbitrary(code);', 'utf8'),
    {
      ...baseTransformOptions,
      dev: false,
      minify: true,
      unstable_transformProfile: 'hermes-canary',
      customTransformOptions: { __proto__: null, bytecode: true },
    }
  );
  expect(result.output[0].data.code).toMatchInlineSnapshot(`
    "__d(function (global, _$$_REQUIRE, _$$_IMPORT_DEFAULT, _$$_IMPORT_ALL, module, exports, _dependencyMap) {
      arbitrary(code);
    });"
  `);
});

it('minifies with Hermes transform profile if bytecode is disabled', async () => {
  const result = await Transformer.transform(
    baseConfig,
    '/root',
    'local/file.js',
    Buffer.from('arbitrary(code);', 'utf8'),
    {
      ...baseTransformOptions,
      dev: false,
      minify: true,
      unstable_transformProfile: 'hermes-canary',
    }
  );
  expect(result.output[0].data.code).toMatchInlineSnapshot(`
    "__d(function (g, r, i, a, m, e, d) {
      minified(code);
    });"
  `);
});

it('counts all line endings correctly', async () => {
  const transformStr = (str: string) =>
    Transformer.transform(baseConfig, '/root', 'local/file.js', Buffer.from(str, 'utf8'), {
      ...baseTransformOptions,
      dev: false,
      minify: false,
    });

  const differentEndingsResult = await transformStr('one\rtwo\r\nthree\nfour\u2028five\u2029six');

  const standardEndingsResult = await transformStr('one\ntwo\nthree\nfour\nfive\nsix');

  expect(differentEndingsResult.output[0].data.lineCount).toEqual(
    standardEndingsResult.output[0].data.lineCount
  );
});

it('outputs comments when `minify: false`', async () => {
  const result = await Transformer.transform(
    baseConfig,
    '/root',
    'local/file.js',
    Buffer.from('/*#__PURE__*/arbitrary(code);', 'utf8'),
    { ...baseTransformOptions, dev: false, minify: false }
  );
  expect(result.output[0].data.code).toMatchInlineSnapshot(`
    "__d(function (global, _$$_REQUIRE, _$$_IMPORT_DEFAULT, _$$_IMPORT_ALL, module, exports, _dependencyMap) {
      /*#__PURE__*/arbitrary(code);
    });"
  `);
});

it('omits comments when `minify: true`', async () => {
  const result = await Transformer.transform(
    baseConfig,
    '/root',
    'local/file.js',
    Buffer.from('/*#__PURE__*/arbitrary(code);', 'utf8'),
    { ...baseTransformOptions, dev: false, minify: true }
  );
  expect(result.output[0].data.code).toMatchInlineSnapshot(`
    "__d(function (g, r, i, a, m, e, d) {
      minified(code);
    });"
  `);
});

it('allows outputting comments when `minify: true`', async () => {
  const result = await Transformer.transform(
    { ...baseConfig, minifierConfig: { output: { comments: true } } },
    '/root',
    'local/file.js',
    Buffer.from('/*#__PURE__*/arbitrary(code);', 'utf8'),
    { ...baseTransformOptions, dev: false, minify: true }
  );
  expect(result.output[0].data.code).toMatchInlineSnapshot(`
    "__d(function (g, r, i, a, m, e, d) {
      /*#__PURE__*/minified(code);
    });"
  `);
});

it('allows the constantFoldingPlugin to not remove used helpers when `dev: false`', async () => {
  // NOTE(kitten): The `constantFoldingPlugin` removes used, inlined Babel helpers, unless
  // the AST path has been re-crawled. If this regressed, check whether `programPath.scope.crawl()`
  // is called before this plugin is run.
  jest.mock('metro-transform-plugins', () => ({
    ...jest.requireActual('metro-transform-plugins'),
    inlinePlugin: () => ({}),
    constantFoldingPlugin: jest.requireActual('metro-transform-plugins').constantFoldingPlugin,
  }));

  const contents = ['import * as test from "test-module";', 'export { test };'].join('\n');

  const result = await Transformer.transform(
    baseConfig,
    '/root',
    'local/file.js',
    Buffer.from(contents, 'utf8'),
    { ...baseTransformOptions, dev: false }
  );
  expect(result.output[0].data.code).toMatchSnapshot();
});
