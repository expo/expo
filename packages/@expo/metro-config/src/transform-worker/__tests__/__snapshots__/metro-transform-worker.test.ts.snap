// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`allows the constantFoldingPlugin to not remove used helpers when \`dev: false\` 1`] = `
"__d(function (global, _$$_REQUIRE, _$$_IMPORT_DEFAULT, _$$_IMPORT_ALL, module, exports, _dependencyMap) {
  Object.defineProperty(exports, "__esModule", {
    value: true
  });
  exports.test = undefined;
  var test = _interopRequireWildcard(_$$_REQUIRE(_dependencyMap[0]));
  exports.test = test;
  function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
  function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && Object.prototype.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
});"
`;

exports[`reports filename when encountering unsupported dynamic dependency 1`] = `"local/file.js:Invalid call at line 3: require(a)"`;

exports[`transforms a module with dependencies 1`] = `
{
  "mappings": "AAA",
  "names": [
    "<global>",
  ],
}
`;

exports[`transforms a simple module 1`] = `
{
  "mappings": "AAA",
  "names": [
    "<global>",
  ],
}
`;

exports[`transforms a simple script 1`] = `
{
  "mappings": "AAA",
  "names": [
    "<global>",
  ],
}
`;

exports[`transforms an es module with asyncToGenerator 1`] = `
"__d(function (global, _$$_REQUIRE, _$$_IMPORT_DEFAULT, _$$_IMPORT_ALL, module, exports, _dependencyMap) {
  var _interopRequireDefault = _$$_REQUIRE(_dependencyMap[0], "@babel/runtime/helpers/interopRequireDefault");
  Object.defineProperty(exports, "__esModule", {
    value: true
  });
  exports.test = test;
  var _asyncToGenerator2 = _interopRequireDefault(_$$_REQUIRE(_dependencyMap[1], "@babel/runtime/helpers/asyncToGenerator"));
  function test() {
    return _test.apply(this, arguments);
  }
  function _test() {
    _test = (0, _asyncToGenerator2.default)(function* () {});
    return _test.apply(this, arguments);
  }
});"
`;

exports[`transforms an es module with asyncToGenerator 2`] = `
{
  "mappings": "AAA,OC",
  "names": [
    "<global>",
    "test",
  ],
}
`;

exports[`transforms async generators 1`] = `
"__d(function (global, _$$_REQUIRE, _$$_IMPORT_DEFAULT, _$$_IMPORT_ALL, module, exports, _dependencyMap) {
  var _interopRequireDefault = _$$_REQUIRE(_dependencyMap[0], "@babel/runtime/helpers/interopRequireDefault");
  Object.defineProperty(exports, "__esModule", {
    value: true
  });
  exports.test = test;
  var _awaitAsyncGenerator2 = _interopRequireDefault(_$$_REQUIRE(_dependencyMap[1], "@babel/runtime/helpers/awaitAsyncGenerator"));
  var _wrapAsyncGenerator2 = _interopRequireDefault(_$$_REQUIRE(_dependencyMap[2], "@babel/runtime/helpers/wrapAsyncGenerator"));
  function test() {
    return _test.apply(this, arguments);
  }
  function _test() {
    _test = (0, _wrapAsyncGenerator2.default)(function* () {
      yield "ok";
    });
    return _test.apply(this, arguments);
  }
});"
`;

exports[`transforms import/export syntax when experimental flag is on 1`] = `
{
  "mappings": "AAA",
  "names": [
    "<global>",
  ],
}
`;
